### PR TITLE
Use Scala Hedgehog in ParseKey test

### PR DESCRIPTION
**Pull request by eed3si9n**
Ref https://github.com/sbt/sbt/issues/4523 

This is a port of ParseKey property-based testing to Scala Hedgehog by the framework's author.
One of the feature is automatic shrinking.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @charleso,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by charleso** [comment]
@eed3si9n Oh wow, didn't think this would go anywhere. Do you mind me asking, what are you intentions regarding hedgehog in SBT with this PR? Would you actually merge this, or is it just to get some more people interested/aware?

[As mentioned on the original issue](https://github.com/sbt/sbt/issues/4523#issuecomment-485220926), I would actually be hesitant about using hedgehog in something critical to the Scala ecosystem without setting some expectations first, and doing some tidying etc.

Thanks!

File: main/src/test/scala/sbt/internal/TestBuild2.scala [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+      Gen.constant { deps =>
+        make(t, deps.toVector)
+      }
+    }
+
+  def genAcyclic[A, T](maxDeps: Range[Int], keyGen: Gen[T], max: Range[Int])(
+      make: T => Gen[Vector[A] => A]
+  ): Gen[Vector[A]] = {
+    if (keyGen == null)
+      println("XX")
```
Comment (charleso) [inline main/src/test/scala/sbt/internal/TestBuild2.scala]: Whoops, my left over debugging, should remove this

**Comment by eed3si9n** [comment]
@charleso 
> Would you actually merge this, or is it just to get some more people interested/aware?

Yes, my intent is to merge this at the earliest point you think we're good to go, and replace the existing test that sometimes fails mysteriously. I was able to improve the test a bit (https://github.com/sbt/sbt/pull/5137/commits/a09197a929652af2650744e75f73b058d12fb724) from looking at some shrunk examples, so I think it's doing some useful things.

File: main/src/test/scala/sbt/internal/TestBuild2.scala [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+import sbt.internal.util.Types.{ const, some }
+import sbt.internal.util.complete.Parser
+import sbt.librarymanagement.Configuration
+
+import java.net.URI
+
+import hedgehog._
+import hedgehog.predef.sequence
+
+object TestBuild2 extends TestBuild2
```
Comment (charleso) [inline main/src/test/scala/sbt/internal/TestBuild2.scala]: Sorry for the bad name. Is it worth renaming? That said, it looks like there aren't many places left that use `TestBuild` (`ParserSpec` and `Delegates`). Would you like me to convert them to hedgehog too if you're going to go ahead with this? I would be happy to do this whenever you like.

**Comment by charleso** [comment]
@eed3si9n Sorry for the random suggestions, how can I best help you on this PR? Did you want me to fork your branch and make the changes there and you can cherry-pick them?

EDIT: My fork branch: https://github.com/eed3si9n/sbt/compare/wip/hedgehog...charleso:wip/hedgehog

**Comment by charleso** [comment]
Just to give you a heads up, and part of my vague reluctance in using hedgehog on a major project, is that I haven't really been paying attention to things like binary (or even source) compatibility. I'm not sure how much you see that being a concern/problem in SBT. I don't think it's necessarily very hard to do, I just haven't had the need up until this point because there doesn't _appear_ to be any other users/consumers. I'm happy to discuss further, I love the idea of hedgehog being used more seriously, but slightly concerned that users will inevitably get frustrated with me/hedgehog because of poor maintenance.

**Comment by eed3si9n** [comment]
> Just to give you a heads up, and part of my vague reluctance in using hedgehog on a major project, is that I haven't really been paying attention to things like binary (or even source) compatibility. I'm not sure how much you see that being a concern/problem in SBT. I don't think it's necessarily very hard to do, I just haven't had the need up until this point because there doesn't appear to be any other users/consumers. I'm happy to discuss further, I love the idea of hedgehog being used more seriously, but slightly concerned that users will inevitably get frustrated with me/hedgehog because of poor maintenance.

In my opinion, you just have to be upfront about whether you intend to keep source- and/or binary- compatibility for whatever range of version number. You already state in the readme that "The API may break during this stage". Now that there is a version number, you can take different strategy about the compatibility.

For example:
- As Semantic Versioning suggests, all 0.x versions are beta versions. API may break between 0.1.0 and 0.2.0.
- We will try to keep source compatibility within 0.x series between the patches. For example, tests written for 0.1.0 should continue to work for 0.1.(1, 2, 3, ...).

sbt cares a lot about the binary compatibility of itself because the plugin ecosystem is one of our best asset, but it's not a huge concern for me if Hedgehog needs to make breaking changes in the future since we're just using to test things. It might eventually be your concern too once people want to integrate with Hedgehog to do law checking (https://github.com/typelevel/discipline) or some other integration test kit off of Hedgehog.

File: main/src/test/scala/sbt/internal/TestBuild2.scala (hunk 2 above)
Comment (eed3si9n) [reply main/src/test/scala/sbt/internal/TestBuild2.scala]: I think if would be cool if we can migrate the other ones so we can remove the old TestBuild.

**Comment by charleso** [comment]
Cool. Should be relatively straight forward. I'm happy to do that in the
next day or so and let you know if you like?

**Comment by eed3si9n** [comment]
FYI the last commit on https://github.com/eed3si9n/sbt/compare/wip/hedgehog...charleso:wip/hedgehog has author information as

```
Charles O'Farrell <cofarrell@Charless-MacBook-Pro.local>
```

This is going to trigger the CLA validator etc, so I amended to match the other commits.

**Comment by charleso** [comment]
@eed3si9n Thanks for fixing my git author.

I've rebased and pushed a new commit to my fork/branch which updates the 2 remaining ScalaCheck tests to hedgehog. 

The last `Delegates` test fails, but at a glance I'm not sure if this is because of any of my/your changes to the generation or not. I'm happy to investigate but I figure you might have a better idea?

```
- sbt.Delegates$.Config axis delegates to parent configuration: Falsified after 20 passed tests
> Env:
    Tasks:
    aa (delegates: )
    ab (delegates: )
Build file:///aaaaaaa/#a :
    Project a
      Delegates:

      Configurations:
        Aa
    Project b
      Delegates:

      Configurations:
        A
Scopes:
    Scope(Select(BuildRef(file:///aaaaaaa/#a)), Select(ConfigKey(A)), Zero, Zero)
> java.util.NoSuchElementException: key not found: A
    at scala.collection.immutable.Map$Map1.apply(Map.scala:114)
    at sbt.internal.TestBuild$Env.inheritConfig(TestBuild.scala:142)
    at sbt.Delegates$.$anonfun$tests$36(Delegates.scala:101)
    at sbt.Delegates$.$anonfun$allDelegates$1(Delegates.scala:124)
    at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
    at scala.collection.immutable.List.foreach(List.scala:392)
    at scala.collection.TraversableLike.map(TraversableLike.scala:238)
    at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
    at scala.collection.immutable.List.map(List.scala:298)
    at sbt.Delegates$.allDelegates(Delegates.scala:122)
    at sbt.Delegates$.$anonfun$tests$35(Delegates.scala:95)
    at scala.Option.map(Option.scala:230)
    at hedgehog.core.PropertyT.$anonfun$map$1(PropertyT.scala:64)
    at scala.Option.map(Option.scala:230)
    at hedgehog.core.GenT.$anonfun$map$2(GenT.scala:12)
    at hedgehog.core.TreeImplicits1$$anon$1.map(Tree.scala:35)
    at hedgehog.core.TreeImplicits1$$anon$1.map(Tree.scala:33)
    at hedgehog.core.Tree.map(Tree.scala:16)
    at hedgehog.core.TreeImplicits1$$anon$1.$anonfun$map$2(Tree.scala:35)
    at hedgehog.predef.LazyList.$anonfun$map$1(LazyList.scala:16)
    at hedgehog.predef.LazyList.$anonfun$map$1(LazyList.scala:16)
    at hedgehog.predef.package$.findMap(package.scala:40)
    at hedgehog.core.PropertyTReporting.takeSmallest(PropertyT.scala:145)
    at hedgehog.core.PropertyTReporting.takeSmallest$(PropertyT.scala:135)
    at hedgehog.package$$anon$2.takeSmallest(package.scala:32)
    at hedgehog.core.PropertyTReporting.loop$1(PropertyT.scala:202)
    at hedgehog.core.PropertyTReporting.report(PropertyT.scala:212)
    at hedgehog.core.PropertyTReporting.report$(PropertyT.scala:169)
    at hedgehog.package$$anon$2.report(package.scala:32)
    at hedgehog.PropertyTOps.check(Property.scala:38)
    at hedgehog.PropertyTOps.check$(Property.scala:37)
    at hedgehog.package$Property$.check(package.scala:19)
    at hedgehog.runner.Properties.$anonfun$main$1(Properties.scala:15)
    at hedgehog.runner.Properties.$anonfun$main$1$adapted(Properties.scala:14)
    at scala.collection.immutable.List.foreach(List.scala:392)
    at hedgehog.runner.Properties.main(Properties.scala:14)
    at sbt.Delegates.main(Delegates.scala)
```

File: main/src/test/scala/ParseKey.scala [hunk 3]
```diff
@@ -7,26 +7,47 @@
...
 import sbt.Def.{ ScopedKey, displayFull, displayMasked }
-import sbt.internal.TestBuild._
-import sbt.internal.util.AttributeKey
-import sbt.internal.util.complete.{ DefaultParsers, Parser }
-import sbt.internal.{ Resolve, TestBuild }
-import sbt.librarymanagement.Configuration
+import sbt.internal.TestBuild2._
+import sbt.internal.util.complete.Parser
+import sbt.internal.{ Resolve, TestBuild2 }
+import hedgehog.{ Result => Assert, _ }
```
Comment (charleso) [inline main/src/test/scala/ParseKey.scala]: @eed3si9n The only reason for the rename is that `sbt` exposes a `Result` object as well. I'm not sure if there's a better/different way to exclude that from `sbt` instead of this rename?

Comment (eed3si9n) [reply main/src/test/scala/ParseKey.scala]: Generally speaking I think putting test cases in `sbt` package might not be a good practice since it gives access to the internals. Sometimes necessary, but I don't know if that should be the default.

Comment (eed3si9n) [reply main/src/test/scala/ParseKey.scala]: Changing to another package didn't really work because we are using stuff from package private..

```
[error] /Users/eed3si9n/work/sbt-modules/sbt/main/src/test/scala/ParseKey.scala:150:37: method fromConfigIdent in trait KeyIndex cannot be accessed in sbt.internal.KeyIndex
[error]       val name = structure.keyIndex.fromConfigIdent(maybeResolvedProj)(configID)
[error]                                     ^
[error] /Users/eed3si9n/work/sbt-modules/sbt/main/src/test/scala/ParseKey.scala:174:29: method displayPedantic in object Scope cannot be accessed in object sbt.Scope
[error]         .log(s"Key: ${Scope.displayPedantic(key.scope, key.key.label)}")
[error]                             ^
```

**Comment by eed3si9n** [comment]
> I've rebased and pushed a new commit to my fork/branch which updates the 2 remaining ScalaCheck tests to hedgehog.

Thanks.

> The last `Delegates` test fails, but at a glance I'm not sure if this is because of any of my/your changes to the generation or not. I'm happy to investigate but I figure you might have a better idea?

Looks like there's a problem with the test.

```
    Scope(Select(BuildRef(file:///aaaaaaa/#a)), Select(ConfigKey(A)), Zero, Zero)
```

`BuildRef(...)` technically doesn't own a configuration, so it's probably failing to look anything up. Filtering BuildRef out for now should be ok - https://github.com/sbt/sbt/pull/5137/commits/1cfe14a877582be8efc5df47dd5faef5bf7a02dc

File: main/src/test/scala/ParseKey.scala (hunk 3 above)
Comment (eed3si9n) [reply main/src/test/scala/ParseKey.scala]: Fully qualified `hedgehog.Result` I think is better than renaming this to `Assert`.

Comment (charleso) [reply main/src/test/scala/ParseKey.scala]: > Fully qualified `hedgehog.Result` I think is better than renaming this to `Assert`

👍 Sounds good. Agreed.

File: main/src/test/scala/sbt/internal/TestBuild.scala [hunk 4]
```diff
@@ -249,54 +253,56 @@ abstract class TestBuild {
...
-      ssp <- nonEmptyId
+      ssp <- pathGen
       frag <- optIDGen
     } yield new URI("file", "///" + ssp + "/", frag.orNull)
   }
 
-  implicit def envGen(implicit bGen: Gen[Build], tasks: Gen[Vector[Taskk]]): Gen[Env] =
-    for (i <- MaxBuildsGen; bs <- containerOfN[Vector, Build](i, bGen); ts <- tasks)
+  def envGen(bGen: Gen[Build], tasks: Gen[Vector[Taskk]]): Gen[Env] =
+    for (bs <- bGen.list(MaxBuildsGen).map(_.toVector); ts <- tasks)
```
Comment (charleso) [inline main/src/test/scala/sbt/internal/TestBuild.scala:291]: @eed3si9n Just a trick I sometimes do to generate unique values in a list, I `zipWithIndex` and then prepend/append to the value that needs to be unique with `_$i` or something like that. Means you don't need to worry about lengths/uniqueness. The only downside is that you're not "testing" things that don't always have `_$i` in that value, which might not test something about the name.

**Comment by charleso** [comment]
@eed3si9n Just to take stock, is there anything else you need/want me to do on this PR? What's next?

**Comment by eed3si9n** [comment]
@charleso I am ok with merging as-is. What do you think?

**Comment by charleso** [comment]
@ee3si9n I'm ok with that if you are? It's exciting (for me at least).

Do let me know if there are any follow questions/concerns/requests for Hedghehog. I'll try to improve my release process and include mima for sanity checking this week now that I have a real user!

Thanks!

**Comment by charleso** [comment]
@eed3si9n thanks so much for all the help, and pushing that through!!!

**Comment by eed3si9n** [comment]
@charleso Thanks for all the help. Hopefully I'd starting writing more tests (none of these tests were originally written by me).
